src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -0,0 +1,257 @@
+      array(end) = null
+      Some(elem.asInstanceOf[A])
+    }
+  }
+
+  override def length = (end - start) & mask
+
+  override def isEmpty = start == end
+
+  override def clone() = new ArrayDeque(array.clone, start, end)
+
+  override def clear() = if (nonEmpty) set(array = ArrayDeque.alloc(ArrayDeque.defaultInitialSize), start = 0, end = 0)
+
+  override def slice(from: Int, until: Int) = {
+    val left = box(from)
+    val right = box(until)
+    val len = right - left
+    if (len <= 0) {
+      ArrayDeque.empty[A]
+    } else if (len >= size) {
+      clone()
+    } else {
+      val array2 = ArrayDeque.alloc(len)
+      arrayCopy(dest = array2, srcStart = left, destStart = 0, maxItems = len)
+      new ArrayDeque(array2, 0, len)
+    }
+  }
+
+  override def sliding(window: Int, step: Int) = {
+    require(window >= 1 && step >= 1, s"size=$size and step=$step, but both must be positive")
+    (indices by step).iterator.map(i => slice(i, i + window))
+  }
+
+  override def grouped(n: Int) = sliding(n, n)
+
+  override def copyToArray[B >: A](dest: Array[B], destStart: Int, len: Int) =
+    arrayCopy(dest = dest, srcStart = 0, destStart = destStart, maxItems = len)
+
+  override def companion = ArrayDeque
+
+  override def result() = this
+
+  override def stringPrefix = "ArrayDeque"
+
+  override def toArray[B >: A: ClassTag] = {
+    val array2 = Array.ofDim[B](size)
+    arrayCopy(dest = array2, srcStart = 0, destStart = 0, maxItems = size)
+    array2
+  }
+
+  def arrayCopy(dest: Array[_], srcStart: Int, destStart: Int, maxItems: Int): Unit = {
+    checkIndex(destStart, dest)
+    checkIndex(srcStart)
+    val toCopy = maxItems min (size - srcStart) min (dest.length - destStart)
+    if (toCopy > 0) {
+      val startIdx = (start + srcStart) & mask
+      val block1 = toCopy min (array.length - startIdx)
+      Array.copy(src = array, srcPos = startIdx, dest = dest, destPos = destStart, length = block1)
+      if (block1 < toCopy) {
+        Array.copy(src = array, srcPos = 0, dest = dest, destPos = block1, length = toCopy - block1)
+      }
+    }
+  }
+
+  /**
+    * Trims the capacity of this CircularBuffer's instance to be the current size
+    */
+  def trimToSize(): Unit = resize(size - 1)
+
+  @inline private def checkIndex(idx: Int, seq: GenSeq[_] = this) = if(!seq.isDefinedAt(idx)) throw new IndexOutOfBoundsException(idx.toString)
+
+  @inline private def box(i: Int) = if (i <= 0) 0 else if (i >= size) size else i
+
+  @inline private def get(idx: Int) = array((start + idx) & mask)
+
+  @inline private def set(idx: Int, elem: AnyRef) = array((start + idx) & mask) = elem
+
+  @inline private def set(array: Array[AnyRef], start: Int, end: Int) = {
+    this.array = array
+    this.mask = array.length - 1
+    assert((array.length & mask) == 0, s"Array.length must be power of 2")
+    this.start = start
+    this.end = end
+  }
+
+  private def ensureCapacity() = if (size == array.length - 1) resize(array.length)
```

Comment:
The issue is that one might call clear() on a large buffer, then add a bunch of elements one at a time after that in order to avoid extra allocation.  In that use case, having it shrink at being below 1/4 full is a bad idea.  Another situation is if the user pre-sizes it to say, 2MB then starts adding things one at a time.  We would not want it to shrink on the first element added in that case either.  Nor would we want it to shrink if we added 100 things, removed 1, added 100, removed 1, etc (e.g. only checking on removal).    It will take a long time before it reaches more than 1/4 full, and it is too hard to know inside of `ensureCapacity()` what the user's intention is.

And since every append / prepend needs to `ensureCapacity()`, it needs to be as simple and fast as possible for the job.

A separate `maybeShrink()` applied selectively might work, but I'm not sure what the right heuristics for that would be -- retain the initial size?  track a recent clear()  call?  That all seems messy to me.

I think if the api did not have a clear() and was purely append/prepend/removeTail/removeHead then auto-shrink at the right time is a better fit.  I would go as far as to say that a structure that auto-shrinks is a different data structure entirely, with a slightly different use case.  I'd rather have users ask for that by type, something like `ShrinkingArrayDeque`.   However, I can't think of a use case where `trimToSize` is not good enough.  Once I am done mutating the buffer, if I intend to hold on to it, I can call that.